### PR TITLE
Fix to Issue #17

### PR DESCRIPTION
**Pull request by kappeld**
This is a fix to issue #17. c-string arguments have been changed to Name object. This was necessary because NEST no longer allows Name objects to be build "on the fly" when multi-threading is turned on. 

In the current commit, there is still a problem with the connection recorder object that needs to be adapted to work with Name object.

### CODE REVIEW

File: examples/pattern_matching_showcase/python/config.py [hunk 1]
```diff
@@ -142,7 +142,7 @@ def s_to_ms(seconds):
 if len(sys.argv) > 1 and "TEST_MODE" in sys.argv[1:]:
     total_simulation_time = 20.0
 
-nest_n_threads = 1  # None -> auto; Integer -> according fixed number of threads
+nest_n_threads = 16  # None -> auto; Integer -> according fixed number of threads
```
Comment (mhoff) [inline examples/pattern_matching_showcase/python/config.py]: I guess we should stick to a lower value here. As a demonstration 2-4 should be totally sufficient.

File: src/spore_names.h [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+ * For more information see: https://github.com/IGITUGraz/spore-nest-module 
+ * 
+ * File:   spore_names.h
+ * Author: Kappel
+ *
+ * Created on June 8, 2017, 5:51 PM
+ */
+
+#ifndef SPORE_NAMES_H
+#define	SPORE_NAMES_H
```
Comment (mhoff) [inline src/spore_names.h]: Indentation?

File: src/sporemodule.cpp [hunk 3]
```diff
@@ -160,7 +160,7 @@ void spore::SporeModule::init(SLIInterpreter* i)
     nest::kernel().model_manager.register_node_model<RewardInProxy>("reward_in_proxy");
 
     ConnectionUpdateManager::instance()->init(cu_model_id);
-
+
```
Comment (mhoff) [inline src/sporemodule.cpp]: To delete.

**Comment by kappeld** [comment]
I have fixed now the problem with the logger. There seems to be still a problem with the reward_in_proxy, since the test stalls. @mhoff do you have time to look at this?

**Comment by kappeld** [comment]
Oh, no. Actually it seems I was just impatient. I tried now to run now the example and everything seems to work now. Can you confirm and double-check everything since I made quite a lot of changes in relatively short time...

**Comment by kappeld** [comment]
@mhoff is it practical to make a unittest with the original example code that caused the bug? Could you prepare that?

File: src/diligent_connector_model.h [hunk 4]
```diff
@@ -274,16 +274,16 @@ nest::ConnectorBase* DiligentConnectorModel< ConnectionT >::cleanup_delete_conne
...
     vector_like< ConnectionT >* vc;
 
-    bool b_has_primary = has_primary(conn);
-    bool b_has_secondary = has_secondary(conn);
+    const bool b_has_primary = has_primary(conn);
+    const bool b_has_secondary = has_secondary(conn);
 
-    conn = validate_pointer(conn);
+    nest::ConnectorBase* conn_vp = validate_pointer(conn);
     // from here on we can use conn as a valid pointer
```
Comment (mhoff) [inline src/diligent_connector_model.h]: `conn` or `conn_vp`?

File: src/spore_names.cpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+const Name parameter_mapping_offset("parameter_mapping_offset");
+const Name weight_scale("weight_scale");
+const Name weight_update_interval("weight_update_interval");
+const Name gradient_scale("gradient_scale");
+const Name dopa_trace_id("dopa_trace_id");
+const Name psp_cutoff_amplitude("psp_cutoff_amplitude");
+const Name simulate_retracted_synapses("simulate_retracted_synapses");
+const Name delete_retracted_synapses("delete_retracted_synapses");
+
+const Name synaptic_parameter("synaptic_parameter");
```
Comment (mhoff) [inline src/spore_names.cpp]: On the long term, we might consider to unify the terminology for `synaptic_parameter`, `max_param`, `min_param` and `max_param_change`. Something like `syn_param`, `max_syn_param`, ... might be more logical.

File: src/synaptic_sampling_rewardgradient_connection.h [hunk 6]
```diff
@@ -555,6 +561,8 @@ ConnectionDataLogger< SynapticSamplingRewardGradientConnection<targetidentifierT
 
 /**
  * Get the data logger singleton.
+ * 
+ * @note Instantiating the logger for the first time may not be thread save.
```
Comment (mhoff) [inline src/synaptic_sampling_rewardgradient_connection.h]: Correction: "thread-safe"

File: src/synaptic_sampling_rewardgradient_connection.h [hunk 7]
```diff
@@ -586,18 +596,29 @@ ConnectionDataLogger< SynapticSamplingRewardGradientConnection<targetidentifierT
...
 
 /**
+ * Check syn_spec dictionary for parameters that are not allowed for this
+ * connection. Will issue warning or throw error if a parameter is found.
+ */
+template <typename targetidentifierT>
+void SynapticSamplingRewardGradientConnection<targetidentifierT>::check_synapse_params(const DictionaryDatum& syn_spec)
+    const
+{
+    // FIXME!! Check synaptic parameters here!
```
Comment (mhoff) [inline src/synaptic_sampling_rewardgradient_connection.h]: Is this a proper TODO? Do we need an issue to keep track of this on the long term?

**Comment by mhoff** [comment]
I can confirm that my current experiment is working just fine with this bug-fix. The version after the first commit also already showed significant learning results after a few hours, thus I can confidently say that these changes should fix the problem.

**Comment by mhoff** [comment]
> @mhoff is it practical to make a unittest with the original example code that caused the bug? Could you prepare that?

Yes. I will take care of that. However, most likely I will not be able to do that until next week. I would be fine with merging this bug-fix branch and adding the test afterwards.

**Comment by kappeld** [comment]
#18 fixes this issue and has been merged into master. I keep this issue open to remind us to add regression tests...

**Pull request merged**